File: packages/node/src/compile.ts
import { execFile } from 'node:child_process';
import { chmod, mkdir, readFile, rename, rm, stat } from 'node:fs/promises';
import { createHash } from 'node:crypto';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';
import { ModelUnavailableError } from './errors.js';
import { findSwiftc, hostTarget } from './target.js';

const execFileAsync = promisify(execFile);

export interface Progress {
  status: string;
  percent?: number;
}
export type OnProgress = (p: Progress) => void;

/**
 * Where the compiled helper is cached.
 *
 * Deliberately not namespaced per language binding: the npm and the pip package
 * embed byte-identical Swift, so they derive the same fingerprint and share one
 * compiled binary. Installing both costs one compile, not two.
 */
export function cacheDir(): string {
  return path.join(os.homedir(), 'Library', 'Caches', 'apple-llm', 'bin');
}

/**
 * The cache key: sha256 over the Swift source, a newline, and the target triple,
 * truncated to 12 hex characters.
 *
 * The exact recipe is load-bearing — the Python package computes the same string
 * and must agree byte for byte, or the two would each compile their own copy.
 * A helper edit or an OS upgrade changes it, so both rebuild automatically.
 */
export function fingerprint(source: string, triple: string): string {
  return createHash('sha256').update(`${source}\n${triple}`, 'utf8').digest('hex').slice(0, 12);
}

let cachedSource: string | undefined;

/** The embedded Swift source, shipped as package data beside the built output. */
export async function helperSource(): Promise<string> {
  if (cachedSource !== undefined) return cachedSource;
  const here = path.dirname(fileURLToPath(import.meta.url));
  // dist/index.js -> ../swift/helper.swift; src/compile.ts -> ../swift/helper.swift
  const candidates = [
    path.join(here, '..', 'swift', 'helper.swift'),
    path.join(here, 'swift', 'helper.swift'),
  ];
  for (const candidate of candidates) {
    try {
      cachedSource = await readFile(candidate, 'utf8');
      return cachedSource;
    } catch {
      // try the next location
    }
  }
  throw new Error(
    `apple-llm could not find its embedded helper.swift (looked in ${candidates.join(', ')}).`,
  );
}

async function fileExists(p: string): Promise<boolean> {
  try {
    await stat(p);
    return true;
  } catch {
    return false;
  }
}

let inFlight: Promise<string> | undefined;
let cachedKey: string | undefined;

/**
 * Compile the helper on first use and cache it. Never called at install time —
 * importing this package on Linux, an Intel Mac or macOS 25 must not fail.
 *
 * `force` bypasses the in-process memo, so a caller can rebuild after the cached
 * binary is removed or found corrupt without restarting the program.
 *
 * The memo is fingerprint-aware: a helper edit or OS upgrade changes the
 * fingerprint, so a stale in-process memo is discarded rather than reused.
 */
export async function ensureBinary(
  onProgress?: OnProgress,
  options: { force?: boolean } = {},
): Promise<string> {
  if (options.force === true) {
    inFlight = undefined;
    cachedKey = undefined;
  } else if (inFlight !== undefined && cachedKey !== undefined) {
    const pending = inFlight;
    try {
      const source = await helperSource();
      const triple = (await hostTarget()) ?? process.arch;
      if (fingerprint(source, triple) === cachedKey) return pending;
      // Fingerprint changed (helper edited or OS upgraded): fall through
      // and rebuild rather than reuse the stale binary.
      inFlight = undefined;
      cachedKey = undefined;
    } catch {
      return pending;
    }
  }
  if (inFlight === undefined) {
    try {
      const source = await helperSource();
      const triple = (await hostTarget()) ?? process.arch;
      cachedKey = fingerprint(source, triple);
    } catch {
      // Fingerprint is best effort; build() recomputes it anyway.
    }
    inFlight = build(onProgress).catch((err) => {
      inFlight = undefined;
      cachedKey = undefined;
      throw err;
    });
  }
  return inFlight;
}

async function build(onProgress?: OnProgress): Promise<string> {
  const source = await helperSource();
  const triple = (await hostTarget()) ?? process.arch;
  const dir = cacheDir();
  const target = path.join(dir, `fm-helper-${fingerprint(source, triple)}`);

  if (await fileExists(target)) return target;

  const swiftc = await findSwiftc();
  if (swiftc === null) {
    throw new ModelUnavailableError(
      'apple-llm needs a Swift compiler to build its on-device helper, but none was found.\n' +
        'Install the Xcode command line tools with `xcode-select --install`, then try again.',
      'noSwiftCompiler',
      'device',
    );
  }

  onProgress?.({ status: 'building the Apple on-device helper (one time, a few seconds)' });
  await mkdir(dir, { recursive: true });
  // Random suffix: same-PID concurrent force:true builds must not collide on
  // one shared temp name.
  const unique = `${process.pid}.${Math.random().toString(36).slice(2)}`;
  const sourcePath = `${target}.${unique}.swift`;
  const staging = `${target}.${unique}.tmp`;
  const { writeFile } = await import('node:fs/promises');
  await writeFile(sourcePath, source, 'utf8');

  try {
    const targetArgs = triple.includes('-apple-') ? ['-target', triple] : [];
    await execFileAsync(swiftc.command, [
      ...swiftc.prefixArgs,
      ...targetArgs,
      '-parse-as-library',
      '-O',
      sourcePath,
      '-o',
      staging,
    ]);
    await chmod(staging, 0o755);
    // Atomic publish, so concurrent first-runs cannot observe a half-written file.
    await rename(staging, target);
  } catch (err) {
    await rm(staging, { force: true });
    // A concurrent run may have finished first, which is a success for us.
    if (await fileExists(target)) return target;
    const reason = err instanceof Error ? err.message : String(err);
    throw new ModelUnavailableError(
      `Failed to build the Apple on-device helper with ${swiftc.command}.\n${reason.slice(0, 400)}\n` +
        'This usually means the macOS SDK predates the Foundation Models framework (macOS 26+).',
      'unsupportedOSVersion',
      'device',
    );
  } finally {
    await rm(sourcePath, { force: true });
  }

  onProgress?.({ status: 'helper built' });
  return target;
}

File: packages/python/src/apple_llm/_compile.py
"""Compile the Swift helper on first use and cache it.

Never called at install time: importing this package on Linux, an Intel Mac or
macOS 25 must succeed, and the wheel is pure Python with no build step.
"""

from __future__ import annotations

import hashlib
import os
import random
import subprocess
import threading
from pathlib import Path
from typing import Callable, Optional

from ._errors import ModelUnavailableError
from ._target import find_swiftc, host_target

#: Called with a short status string for the one-time compile and the shortcut
#: install. Both take seconds and must not look like a hang.
ProgressCallback = Callable[[str], None]

_lock = threading.Lock()
_binary_path: Optional[str] = None
_binary_fingerprint: Optional[str] = None
_source_cache: Optional[str] = None


def cache_dir() -> Path:
    """Where the compiled helper is cached.

    Deliberately not namespaced per language binding: the npm and the pip package
    embed byte-identical Swift, so they derive the same fingerprint and share one
    compiled binary. Installing both costs one compile, not two.
    """
    return Path.home() / "Library" / "Caches" / "apple-llm" / "bin"


def fingerprint(source: str, triple: str) -> str:
    """The cache key: sha256 over the source, a newline, and the target triple,
    truncated to 12 hex characters.

    The exact recipe is load-bearing -- the Node package computes the same string
    and must agree byte for byte, or the two would each compile their own copy.
    A helper edit or an OS upgrade changes it, so both rebuild automatically.
    """
    digest = hashlib.sha256(f"{source}\n{triple}".encode()).hexdigest()
    return digest[:12]


def helper_source() -> str:
    """The embedded Swift source, shipped as package data."""
    global _source_cache
    if _source_cache is None:
        _source_cache = (Path(__file__).parent / "swift" / "helper.swift").read_text(
            encoding="utf-8"
        )
    return _source_cache


def ensure_binary(on_progress: Optional[ProgressCallback] = None, force: bool = False) -> str:
    """Return the path to the compiled helper, building it if needed."""
    global _binary_path, _binary_fingerprint
    with _lock:
        source = helper_source()
        triple = host_target() or ""
        fp = fingerprint(source, triple)
        if (
            _binary_path is not None
            and _binary_fingerprint == fp
            and not force
            and os.path.exists(_binary_path)
        ):
            return _binary_path
        _binary_path = _build(on_progress)
        _binary_fingerprint = fp
        return _binary_path


def _build(on_progress: Optional[ProgressCallback]) -> str:
    source = helper_source()
    triple = host_target() or ""
    directory = cache_dir()
    target = directory / f"fm-helper-{fingerprint(source, triple)}"

    if target.exists():
        return str(target)

    swiftc = find_swiftc()
    if swiftc is None:
        raise ModelUnavailableError(
            "apple-llm needs a Swift compiler to build its on-device helper, but none "
            "was found.\nInstall the Xcode command line tools with `xcode-select "
            "--install`, then try again.",
            "noSwiftCompiler",
            "device",
        )

    if on_progress:
        on_progress("building the Apple on-device helper (one time, a few seconds)")
    directory.mkdir(parents=True, exist_ok=True)
    unique = f"{os.getpid()}.{threading.get_ident()}.{random.randrange(1 << 30)}"
    source_path = directory / f"{target.name}.{unique}.swift"
    staging = directory / f"{target.name}.{unique}.tmp"
    source_path.write_text(source, encoding="utf-8")

    command, prefix = swiftc
    target_args = ["-target", triple] if "-apple-" in triple else []
    try:
        subprocess.run(
            [command, *prefix, *target_args, "-parse-as-library", "-O", str(source_path),
             "-o", str(staging)],
            capture_output=True,
            text=True,
            timeout=300,
            check=True,
        )
        staging.chmod(0o755)
        # Atomic publish, so concurrent first-runs cannot observe a half-written
        # file. os.replace is atomic within a filesystem.
        os.replace(staging, target)
    except (OSError, subprocess.SubprocessError) as err:
        staging.unlink(missing_ok=True)
        # A concurrent run may have finished first, which is a success for us.
        if target.exists():
            return str(target)
        detail = getattr(err, "stderr", None) or str(err)
        raise ModelUnavailableError(
            f"Failed to build the Apple on-device helper with {command}.\n{str(detail)[:400]}\n"
            "This usually means the macOS SDK predates the Foundation Models framework "
            "(macOS 26+).",
            "unsupportedOSVersion",
            "device",
        ) from err
    finally:
        source_path.unlink(missing_ok=True)

    if on_progress:
        on_progress("helper built")
    return str(target)
